### PR TITLE
chore: Add generics to typescript generator

### PR DESCRIPTION
**Pull request by Emyrk**
# Golang

```golang
type Flaggable interface {
	string | bool | int | time.Duration | []string
}

type SomethingElse interface {
	time.Duration | bool
}

type DeploymentConfigField[T Flaggable, S SomethingElse] struct {
	Key        string
	Name       string
	Usage      string
	Flag       string
	Shorthand  string
	Enterprise bool
	Hidden     bool
	Value      T
	Else       S
}
```

# Typescript

I think this is correct typescript

```typescript

// From codersdk/deploymentconfig.go
export interface DeploymentConfigField<T extends Flaggable, S extends SomethingElse> {
  readonly Key: string
  readonly Name: string
  readonly Usage: string
  readonly Flag: string
  readonly Shorthand: string
  readonly Enterprise: boolean
  readonly Hidden: boolean
  readonly Value: T
  readonly Else: S
}

// From codersdk/deploymentconfig.go
export type Flaggable = string | boolean | number | string[]

// From codersdk/deploymentconfig.go
export type SomethingElse = number | boolean
```

### CODE REVIEW

File: scripts/apitypings/main.go [hunk 1]
```diff
@@ -253,10 +296,62 @@ func (g *Generator) posLine(obj types.Object) string {
...
+		allTypes = append(allTypes, scriptType.ValueType)
+		optional = optional || scriptType.Optional
+	}
+
+	qMark := ""
+	if optional {
+		qMark = "?"
+	}
+
+	s.WriteString(fmt.Sprintf("export type %s%s = %s\n", obj.Name(), qMark, strings.Join(allTypes, " | ")))
```
Comment (deansheather) [inline scripts/apitypings/main.go:320]: I don't think question marks are supported on types, only on interface fields.

**Review by deansheather — APPROVED** [review]

File: scripts/apitypings/main.go (hunk 1 above)
Comment (Emyrk) [reply scripts/apitypings/main.go:320]: I have read the CLA Document and I hereby sign the CLA

**Review by Emyrk — COMMENTED** [review]
I have read the CLA Document and I hereby sign the CLA

**Comment by Emyrk** [comment]
I have read the CLA Document and I hereby sign the CLA

File: scripts/apitypings/main.go (hunk 1 above)
Comment (Emyrk) [reply scripts/apitypings/main.go]: Good catch. Made it nullable

**Comment by Emyrk** [comment]
Moved to `main` https://github.com/coder/coder/pull/4664

**Pull request closed without merging**